### PR TITLE
migrate bodycomplex in autorest.testserver (for complex polymorphism and dictionary)

### PR DESCRIPTION
**Pull request by changlong-liu**
This PR is for migrating [body-complex](https://github.com/Azure/autorest.testserver/blob/main/swagger/body-complex.json) in autorest.testserver into cadl.
Since there are already projects for primitives, collections and readonly in cadl-ranch, I don't add related tests in body-complex to here.
I split the remains into two packages as `dictionary` and `inheritance-complex`:
- dictionary: for cadl datatype `Record`
- inheritance-complex: provide testcases for complex polymorphic scenario

# Cadl Ranch Contribution Checklist:

- [ ] I have written a [scenario spec](../docs/writing-scenario-spec.md)
- [ ] I have **meaningful** `@scenario` names. Someone can look at the list of scenarios and understand what I'm covering.
- [ ] I have written a [mock API](../docs/writing-mock-apis.md)
- [ ] I have used `@scenarioDoc`s for extra scenario description and to tell people **how to pass** my mock api check.

### CODE REVIEW

File: packages/cadl-ranch-specs/http/models/dictionary/main.cadl [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+using Cadl.Http;
+using Azure.Core.Foundations;
+
+@serviceTitle("Dictionary in Complex Test Service")
+@serviceVersion("2016-02-29")
+@server("http://localhost:3000", "Testserver endpoint")
+@doc("Illustrates complex model generation, as well as serialization and deserialization.")
+namespace Cadl.Testserver.BodyComplexDictionary;
+
+model dictionary_wrapper {
```
Comment (msyyc) [inline packages/cadl-ranch-specs/http/models/dictionary/main.cadl:13]: Model name shall be camel-case

Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/models/dictionary/main.cadl]: changed to `dictionaryWrapper`

File: packages/cadl-ranch-specs/http/models/dictionary/main.cadl [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import "@azure-tools/cadl-azure-core";
+import "@azure-tools/cadl-ranch-expect";
+
+using Cadl.Http;
+using Azure.Core.Foundations;
+
+@serviceTitle("Dictionary in Complex Test Service")
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/models/dictionary/main.cadl:7]: For cadl-ranch we want to be sure that every scenario is valid in itself. You can use `body complex` as inspiration to find missing scenarios in the testserver, but the test shouldn't be tied to body complex in any way

File: packages/cadl-ranch-specs/http/models/dictionary/main.cadl [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+@serviceVersion("2016-02-29")
+@server("http://localhost:3000", "Testserver endpoint")
+@doc("Illustrates dictionary generation, as well as serialization and deserialization.")
+namespace Cadl.Testserver.BodyComplexDictionary;
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/models/dictionary/main.cadl:11]: `Models.Dictionary`

File: packages/cadl-ranch-specs/http/models/dictionary/main.cadl [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+model dictionaryWrapper {
+  defaultProgram?: Record<string>;
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/models/dictionary/main.cadl:14]: `defaultProgram` isn't a good property name, we can do `optionalDictionaryProperty` (and maybe even add a `requiredDictionaryProperty`.

File: packages/cadl-ranch-specs/http/models/dictionary/main.cadl [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+@scenario
+@route("/complex/dictionary/typed")
+namespace dictionary {
+  @route("/valid")
+  @scenarioDoc("Put complex types with dictionary properties")
+  @put
+  op putValid(
+    @doc("Please put a dictionary with 5 key-value pairs: \"txt\":\"notepad\", \"bmp\":\"mspaint\", \"xls\":\"excel\", \"exe\":\"\", \"\":null")
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/models/dictionary/main.cadl:24]: The `@doc` property is more for describing the property, i.e. what we would generate if this was a client library. The description of what the mock api tests should expect should be in the `@scenarioDoc` header. See the docs for more info https://github.com/Azure/cadl-ranch/blob/main/docs/writing-scenario-spec.md

**Review by iscai-msft — COMMENTED** [review]
@changlong-liu can you read over the [docs](https://github.com/Azure/cadl-ranch/tree/main/docs) and align your tests so they aren't tied to `BodyComplex`, and every scenario is clearly documented and necessary? Thanks!

**Comment by changlong-liu** [comment]
> @changlong-liu can you read over the [docs](https://github.com/Azure/cadl-ranch/tree/main/docs) and align your tests so they aren't tied to `BodyComplex`, and every scenario is clearly documented and necessary? Thanks!

Hi @iscai-msft ,  thanks for the information!
I refactored the scenarios and mockapi to align the naming and doc style, and make each scenario for single purpose. The mock apis in PR can get work in my local Python tests.

File: packages/cadl-ranch-specs/http/models/dictionary-basic/main.cadl [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import "@azure-tools/cadl-azure-core";
+import "@azure-tools/cadl-ranch-expect";
+
+using Cadl.Http;
+using Azure.Core.Foundations;
+
+@serviceTitle("DictionaryPropertiesBasicClient")
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/models/dictionary-basic/main.cadl:7]: we're against using qualifiers like "basic" and "complex". I think the "models/dictionary-basic" code shouldn't be its own test (since we're testing dictionary properties). Instead, it should be added to https://github.com/Azure/cadl-ranch/tree/main/packages/cadl-ranch-specs/http/models/property-types.

Comment (iscai-msft) [reply packages/cadl-ranch-specs/http/models/dictionary-basic/main.cadl:7]: And I just want to clarify that just because it's in the testserver means we have to have it in cadl ranch. Looking at these tests, we don't need separate tests for a dictionary property with values and without, because this testing will be captured by a combination of `optional-properties` and `property-types` tests

File: packages/cadl-ranch-specs/http/models/inheritance-complex/main.cadl [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+@doc("The second level model in three levels inheritance with discriminator and defined a new discriminator.")
+@discriminator("discriminator2")
+model SubModelWithDiscriminator extends BaseModelWithDiscriminator {
+  discriminator1: "SubModel";
+}
+
+@doc("The second level model in three levels inheritance with discriminator.")
+model DerivedFromBaseModelWithDiscriminator extends BaseModelWithDiscriminator {
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/models/inheritance-complex/main.cadl:43]: I understand that you're following the naming schema of the existing `inheritance` models, but I've been meaning to change that one as well, since it's not clear what the models are. @timotheeguerin I honestly think in situations like these, it's more clear to name these models things like `Animal`, `Pet`, `Dog`, `Puppy` for example, since you can easily visualize the relationship between these models. The scenario names shouldn't include references to dogs etc, but I think it makes sense for the models to have a metaphorical relationship. What are your thoughts?

File: packages/cadl-ranch-specs/http/models/special-characters/main.cadl [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+using Azure.Core.Foundations;
+
+@serviceTitle("SpecialCharacterClient")
+@serviceVersion("1.0.0")
+@server("http://localhost:3000", "TestServer endpoint")
+@doc("Illustrates models with special characters or keyword in property name")
+@route("/special-chars")
+namespace Model.SpecialCharacter;
+
+@doc("Base type with a discriminator containing dot.")
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/models/special-characters/main.cadl:15]: I don't think `special-characters` should be tied to models. We can have a separate issue to add `special-characters` tests directly under `http`. This way we can test special characters for mdoels, operations, etc

**Review by iscai-msft — COMMENTED** [review]
looking better!

File: packages/cadl-ranch-specs/http/models/inheritance-complex/main.cadl (hunk 7 above)
Comment (timotheeguerin) [reply packages/cadl-ranch-specs/http/models/inheritance-complex/main.cadl:43]: yeah the models can have those names, just the scenarios names should be clear about whwat they do. The implementation details of a scenario(like model names) should be tailored to make the life easy for generator looking to implement the scenario

File: packages/cadl-ranch-specs/http/README.md [hunk 9]
```diff
@@ -25,3 +25,5 @@ The tests in the http/models directory describe scenarios for generating models
 - [nested-models](https://github.com/Azure/cadl-ranch/tree/main/packages/cadl-ranch-specs/http/models/nested-models) - Generate, send, and receive input, output, and round-trip models with required nested model properties.
 - [readonly-properties](https://github.com/Azure/cadl-ranch/tree/main/packages/cadl-ranch-specs/http/models/readonly-properties) - Generate, send, and receive input, output, and round-trip models with readonly properties.
 - [inheritance](https://github.com/Azure/cadl-ranch/tree/main/packages/cadl-ranch-specs/http/models/inheritance) - Generate, send, and receive round-trip models polymorphic models.
+- [inheritance-complex](https://github.com/Azure/cadl-ranch/tree/main/packages/cadl-ranch-specs/http/models/inheritance-complex) - Generate, send, and receive complex polymorphic models.
```
[the remark this answers is not in the document]
Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/README.md]: I moved them to the `inheritance`.
Since the tests in original `inheritance-complex` can fully cover the existing `inheritance`, I think it's fine to remove the existing model&operations in the heritance?

File: packages/cadl-ranch-specs/http/models/dictionary-basic/main.cadl (hunk 6 above)
Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/models/dictionary-basic/main.cadl]: moved dictionary test into `property-types`

File: packages/cadl-ranch-specs/http/models/special-characters/main.cadl (hunk 8 above)
Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/models/special-characters/main.cadl]: removed `special-characters` from this PR

File: packages/cadl-ranch-specs/http/models/inheritance-complex/main.cadl (hunk 7 above)
Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/models/inheritance-complex/main.cadl]: renamed model&property names to make the model more easy to understand

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl [hunk 10]
```diff
@@ -1,60 +1,112 @@
...
 import "@azure-tools/cadl-ranch-expect";
 
 using Cadl.Rest;
 using Cadl.Http;
 using Azure.Core.Foundations;
 
-@doc("Illustrates round-trip polymorphic models.")
-@scenarioService("/polymorphic")
-namespace BasicPolymorphicModels;
+@serviceTitle("ComplexInheritanceClient")
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/models/inheritance/main.cadl:8]: You can do `@scenarioService("/models/inheritance")` and just have doc, scenarioService, and namespace

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl [hunk 11]
```diff
@@ -1,60 +1,112 @@
...
+@route("/inheritance")
+namespace Model.Inheritance;
 
-@doc("Example base type.")
-model BaseClass {
-  @doc("An example property.")
-  baseClassProperty: string;
+@doc("Base model for non-discriminator three levels inheritance.")
+model Pet {
+  name?: string;
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/models/inheritance/main.cadl:17]: I think `name` can be a required property

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl [hunk 12]
```diff
@@ -1,60 +1,112 @@
...
 }
 
-model DerivedFromBaseClassA extends BaseClass {
-  @doc("An example property on a derived type")
-  derivedClassAProperty: string;
+@doc("The second level model A in the non-discriminator three levels inheritance.")
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/models/inheritance/main.cadl:20]: Update doc, and I think `age` can just be required. To be honest most of these properties can be required unless you're specifically trying to test optionality

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl [hunk 13]
```diff
@@ -1,60 +1,112 @@
...
+  sharktype: "goblin";
 }
 
 @scenario
-@scenarioDoc("Generate, send, and receive round-trip inherited model.")
-@route("/model")
+@scenarioDoc("Generate and send model. The valid input value is {name: \"abc\", age: 32, smart: true}")
+@route("/valid")
+@post
+op sendInheritanceValid(@body input: Siamese): OkResponse | ErrorResponse;
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/models/inheritance/main.cadl:64]: nit: you don't need to repeat `inheritance` in the name, you can just say `putValid` and `getValid` etc

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl (hunk 11 above)
Comment (iscai-msft) [reply packages/cadl-ranch-specs/http/models/inheritance/main.cadl:17]: I'd also specify in the doc that this is normal multiple level inheritance, then specify in the doc for the polymorphic model that its for polymorphic inheritance

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl [hunk 14]
```diff
@@ -1,60 +1,112 @@
...
-@scenarioDoc("Generate, send, and receive round-trip model with a polymorphic property.")
-@route("/property")
+@route("/invalid")
+@scenarioDoc("Get a basic model that is invalid for the local strong kind.")
+@get
+op getInheritanceInvalid(): Pet | ErrorResponse;
+
+@scenario
+@scenarioDoc("Generate, send, and receive round-trip model that optional property is not presented.")
+@route("/empty")
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/models/inheritance/main.cadl:86]: I'm not convinced we need multiple dedicated tests for invalid and empty, @timotheeguerin thoughts?

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl [hunk 15]
```diff
@@ -1,60 +1,112 @@
...
-op setValueWithPolymorphicProperty(
-  @body input: ModelWithPolymorphicProperty
-): ModelWithPolymorphicProperty | ErrorResponse;
+op setInheritanceEmpty(@body input: Siamese): Pet | ErrorResponse;
+
+@scenario
+@scenarioDoc("Get a basic model that optional property is sent as null.")
+@route("/null")
+@get
+op getInheritanceWithNullProperty(): Pet | ErrorResponse;
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/models/inheritance/main.cadl:94]: what are we trying to test here? Sorry I'm just trying to figure out what scenario we're trying to test and if it's a good scenario to test

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl [hunk 16]
```diff
@@ -1,60 +1,112 @@
...
+
+@scenario
+@route("/polymorphism/valid")
+@scenarioDoc("Generate, send, and receive round-trip models in three levels inheritance with 2 discriminators. The valid input is  { age: 1, kind: \"shark\",  sharktype: \"goblin\"}")
+@put
+op setPolymorphism(@body input: Fish): Fish | ErrorResponse;
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/models/inheritance/main.cadl:100]: I think we can do `getPolymorphicModel` and `putPolymorphicModel`, one just returning the model and one putting it

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl [hunk 17]
```diff
@@ -1,60 +1,112 @@
...
+
+@scenario
+@route("/polymorphism/recursive")
+@scenarioDoc("Generate, send, and receive round-trip models has collection and dictionary properties referring to models in three levels inheritance with 2 discriminator")
+@put
+op setPolymorphismRecursively(@body input: Salmon): Salmon | ErrorResponse;
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/models/inheritance/main.cadl:106]: I think for polymorphic models we always use the base class, is that correct @timotheeguerin

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl (hunk 10 above)
Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/models/inheritance/main.cadl]: modified

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl (hunk 13 above)
Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/models/inheritance/main.cadl]: done

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl (hunk 14 above)
Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/models/inheritance/main.cadl]: These abnormal cases are found in bodycomplex :), below are my understanding:
- The `invalid` is for testing what happen when service return an incorrect type of property value, for example: return an integer value for a string property.
- The `null` is for testing when service return a null property explicitly.
- The `empty` is for testing optional property is not sent. (The service returns null property implicitly)

Since they do not be related to inheritance and I find the `empty` has just been covered in `property-optional`, I remove them from this PR. And trying to create new PR to add `null` test in `property-optional`.

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl (hunk 15 above)
Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/models/inheritance/main.cadl]: It's for testing what happen when service return an optional property as null.
Moved this test case to `property-optional``in this PR https://github.com/Azure/cadl-ranch/pull/92

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl (hunk 16 above)
Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/models/inheritance/main.cadl]: done

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl (hunk 17 above)
Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/models/inheritance/main.cadl]: Originally I use `Salmon` because the mockapi input/output an Salmon. But yes, base class `Fish` make more sense as an operation body.
Have changed to `Fish`.

File: packages/cadl-ranch-specs/http/models/property-types/main.cadl [hunk 18]
```diff
@@ -146,3 +146,22 @@ model CollectionsModelProperty is ModelTemplate<InnerModel[]>;
 @route("/collections/model")
 interface CollectionsModel
   extends ModelOperations<CollectionsModelProperty, "[{'property': 'hello'}, {'property': 'world'}]"> {}
+
+// Test a model with a string dictionary property
```
Comment (timotheeguerin) [inline packages/cadl-ranch-specs/http/models/property-types/main.cadl:150]: I don't think this should be added here. Point of this spec is to make sure all the different types works as the type of a property.(check an `int` work, check a `dictionary` work, etc.) It shouldn't go in details. For that we should have a dedicated spec testing dictionaries

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl [hunk 19]
```diff
@@ -1,60 +1,253 @@
...
 using Cadl.Rest;
 using Cadl.Http;
 using Azure.Core.Foundations;
 
-@doc("Illustrates round-trip polymorphic models.")
-@scenarioService("/polymorphic")
-namespace BasicPolymorphicModels;
+@doc("Illustrates inheritance and polymorphic model.")
+@scenarioService("/models/inheritance")
+namespace Models.Inheritance;
```
Comment (timotheeguerin) [inline packages/cadl-ranch-specs/http/models/inheritance/main.cadl:10]: This is great name! specially as there is also discriminated union in cadl that we'll need to support which will difer from this but is also polymorhism

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl [hunk 20]
```diff
@@ -1,60 +1,253 @@
...
+@scenarioDoc("""
+Generate and receive model.
+Expected response body:
+```json
+{"name": "abc", "age": 32, "smart": true}
+```
+""")
+@route("/valid")
+@get
+op getValid(): Siamese | ErrorResponse;
```
Comment (timotheeguerin) [inline packages/cadl-ranch-specs/http/models/inheritance/main.cadl:79]: I would remove `| ErrorResponse` from all the operations. We should not care about errors in most specs.( if there is an error, it is unexpected and the test fail)

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl [hunk 21]
```diff
@@ -1,60 +1,253 @@
...
+@route("/polymorphism/valid")
+@scenarioDoc("""
+Generate and receive polymorphic model in multiple levels inheritance with 2 discriminators.
+Expected response body:
+```json
+{"age": 1, "kind": "shark",  "sharktype": "goblin"}
+```
+""")
+@get
+op getPolymorphicModel(): Fish | ErrorResponse;
```
Comment (timotheeguerin) [inline packages/cadl-ranch-specs/http/models/inheritance/main.cadl:97]: Should the `polymorphism` operations be moved inside a subnamespace called `Polymorphism` or maybe `Discriminated` and change the route as well

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl [hunk 22]
```diff
@@ -1,60 +1,253 @@
...
-): ModelWithPolymorphicProperty | ErrorResponse;
+op putRecursivePolymorphicModel(@body input: Fish): OkResponse | ErrorResponse;
+
+@scenario
+@route("/polymorphism/missingdiscriminator")
+@scenarioDoc("""
+Get a model omitting the discriminator.
+""")
+@get
+op getPolymorphismMissingDiscriminator(): Fish | ErrorResponse;
```
Comment (timotheeguerin) [inline packages/cadl-ranch-specs/http/models/inheritance/main.cadl:253]: Should we have one where the discriminator is unknown as well?

**Review by timotheeguerin — COMMENTED** [review]
I think the inheritance spec looks pretty good overall, just a few organizing issues. 

For the null case, I think we should maybe have a dedicated tests for passing null somewhere else because it isn't tied to optional properties.

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl (hunk 21 above)
Comment (iscai-msft) [reply packages/cadl-ranch-specs/http/models/inheritance/main.cadl:97]: yeah i like the idea of a `polymorphism` subnamespace

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl (hunk 22 above)
Comment (iscai-msft) [reply packages/cadl-ranch-specs/http/models/inheritance/main.cadl:253]: oh good idea!

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl (hunk 21 above)
Comment (timotheeguerin) [reply packages/cadl-ranch-specs/http/models/inheritance/main.cadl:97]: I think `Discriminated` make more sense as technically with inheritance in nominal language it is always polymorphism

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl (hunk 20 above)
Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/models/inheritance/main.cadl]: removed all `| ErrorResponse` since there is no error tests

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl (hunk 21 above)
Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/models/inheritance/main.cadl]: moved polymorphic tests into interface `Discriminated` and routes are updated as well

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl (hunk 22 above)
Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/models/inheritance/main.cadl]: yes, added a `getWrongDiscriminator` :)

File: packages/cadl-ranch-specs/http/models/property-types/main.cadl (hunk 18 above)
Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/models/property-types/main.cadl:150]: Hi @timotheeguerin , cadl `Record<>` type is not covered in existing `property-types/main.cadl`, so I'm trying to add a basic test for it. 

Just want to check `dictionary` can work.
Is it valuable to keep at least one `Record<>` test in this `property-types`?

Comment (timotheeguerin) [reply packages/cadl-ranch-specs/http/models/property-types/main.cadl:150]: Yeah I think is is valuable to add a single test here to make sure properties and dictionaries work together.

Comment (timotheeguerin) [reply packages/cadl-ranch-specs/http/models/property-types/main.cadl:150]: and then in `dictionaries/main.cadl` we can have the more in depth test for dictionaries whatever they are

File: packages/cadl-ranch-specs/http/dictionary/main.cadl [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+    {
+      doc: TDoc,
+    }
+  )
+  @put
+  put(@body body: TModel): OkResponse;
+}
+
+// Test a model with an int dictionary property
+@doc("Model with dictionary int properties")
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/dictionary/main.cadl:50]: @timotheeguerin correct me if I'm wrong, but I think for the dictionary tests, we really want to be focusing on testing dictionaries themselves, not dictionaries as properties. We can add dictionary property tests to the `property-types` test, but for the dictionary test we don't want them all to be tied to models

File: packages/cadl-ranch-specs/http/models/inheritance/main.cadl [hunk 24]
```diff
@@ -4,53 +4,268 @@ import "@azure-tools/cadl-ranch-expect";
...
+        ]
+      }
+    }
+  }
+  ```
+  """)
+  @put
+  putRecursiveModel(@body input: Fish): OkResponse;
+
+  @scenario
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/models/inheritance/main.cadl:248]: inheritance looks good to me!

File: packages/cadl-ranch-specs/http/dictionary/main.cadl [hunk 25]
```diff
@@ -0,0 +1,99 @@
+import "@cadl-lang/rest";
+import "@azure-tools/cadl-ranch-expect";
+
+@doc("Illustrates various of dictionary.")
+@scenarioService("/dictionary")
+namespace Dictionary;
+using Cadl.Http;
```
Comment (timotheeguerin) [inline packages/cadl-ranch-specs/http/dictionary/main.cadl:7]: nit: would move using above the file namespace.

File: packages/cadl-ranch-specs/http/dictionary/main.cadl [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+@scenarioDoc("""
+Dictionary as output in body.
+Expected output body:
+```json
+{ "k1": "hello", k2: "world" }
+```
+""")
+@route("/body")
+@get
+op getBody(): {
```
Comment (timotheeguerin) [inline packages/cadl-ranch-specs/http/dictionary/main.cadl:84]: The other test are great but I think we just don't need to test this as a property but for every test we can just get or put the Record directly. (No need for this explicit get/putBody test as every test should already be using the record at the root of the body)

File: packages/cadl-ranch-specs/http/dictionary/main.cadl [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+
+// TEMPLATES
+@doc("Template type for testing dictionary with specific value type. Pass in the type of the value you are looking for")
+model ModelValueTemplate<TProperty> {
+  @doc("Property")
+  property: Record<TProperty>;
+}
+
+@doc("Template to have models operations")
+interface ModelOperations<TModel, TDoc extends string> {
```
[the remark this answers is not in the document]
Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/dictionary/main.cadl]: yes, changed the naming in new PR

**Comment by changlong-liu** [comment]
Hi @timotheeguerin , @iscai-msft , I moved the dictionary changes into PR https://github.com/Azure/cadl-ranch/pull/104 and leave this PR only for inheritance :)

File: packages/cadl-ranch-specs/http/dictionary/main.cadl (hunk 26 above)
Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/dictionary/main.cadl]: yes, I put dictionary as root type in new PR

File: packages/cadl-ranch-specs/http/dictionary/main.cadl [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+using Cadl.Http;
+
+// TEMPLATES
+@doc("Template type for testing dictionary with specific value type. Pass in the type of the value you are looking for")
+model ModelValueTemplate<TProperty> {
+  @doc("Property")
+  property: Record<TProperty>;
+}
+
+@doc("Template to have models operations")
```
[the remark this answers is not in the document]
Comment (changlong-liu) [reply packages/cadl-ranch-specs/http/dictionary/main.cadl]: learnt from the `property-types` :)

**Review by timotheeguerin — APPROVED** [review]
sweet! looks great!
